src/clib/pio_nc.c

```diff
@@ -2278,24 +2342,54 @@ int PIOc_def_dim(int ncid, const char *name, PIO_Offset len, int *idp)
     }
 
     /* ADIOS: assume all procs are also IO tasks */
-#ifdef _ADIOS
+#ifdef _ADIOS2
     if (file->iotype == PIO_IOTYPE_ADIOS)
     {
         LOG((2, "ADIOS define dimension %s with size %llu, id = %d",
                 name, (unsigned long long)len, file->num_dim_vars));
 
         char dimname[PIO_MAX_NAME];
-        snprintf(dimname, sizeof(dimname), "/__pio__/dim/%s", name);
-        adios_define_var(file->adios_group, dimname, "", adios_unsigned_long, "", "", "");
+        snprintf(dimname, PIO_MAX_NAME, "/__pio__/dim/%s", name);
+        adios2_variable *variableH = adios2_inquire_variable(file->ioH, dimname);
+        if (variableH == NULL)
+        {
+            variableH = adios2_define_variable(file->ioH, dimname, adios2_type_uint64_t,
+                                               0, NULL, NULL, NULL,
+                                               adios2_constant_dims_false);
+            if (variableH == NULL)
+            {
+                return pio_err(ios, file, PIO_EADIOS2ERR, __FILE__, __LINE__, "Defining (ADIOS) variable (name=%s) failed for file (%s, ncid=%d)", dimname, pio_get_fname_from_file(file), file->pio_ncid);
+            }
+        }
 
         assert(file->num_dim_vars < PIO_MAX_DIMS);
         file->dim_names[file->num_dim_vars] = strdup(name);
         file->dim_values[file->num_dim_vars] = len;
         *idp = file->num_dim_vars;
         ++file->num_dim_vars;
-        adios_write(file->adios_fh, dimname, &len);
+        adios2_error adiosErr = adios2_put(file->engineH, variableH, &len, adios2_mode_sync);
+        if (adiosErr != adios2_error_none)
+        {
+            return pio_err(ios, file, ierr, __FILE__, __LINE__, "adios2_put failed, error code = %d", adiosErr);
+        }
 
-        ierr = 0;
+        ierr = PIO_NOERR;
+
+        if (len == PIO_UNLIMITED)
+        {
+            file->num_unlim_dimids++;
+            file->unlim_dimids = (int *)realloc(file->unlim_dimids,
+                                                file->num_unlim_dimids * sizeof(int));
+            if (!file->unlim_dimids)
+            {
+                return pio_err(ios, file, PIO_ENOMEM, __FILE__, __LINE__, "Error: %d", 1);
```

Comment:
This error message does not look right

Reply:
That is inherited from adios2_support branch, thanks for finding it out.